File: boj/b_17298.js
function solution(n, numbers) {
  const result = Array(n).fill(-1);
  const s = [];
  numbers.forEach((number, i) => {
    while (s.length && s[s.length - 1][0] < number) {
      const [v, idx] = s.pop();
      result[idx] = number;
    }
    s.push([number, i]);
  });
  return result;
}
console.log(solution(5, [7, 8, 10, 12, 1]));

File: boj/b_17298.py
import sys
sys.stdin = open("input.txt", "r") # 제거
input = sys.stdin.readline

n = int(input())
table = list(map(int, input().split()))
s = []
result = [-1 for _ in range(n)]
for i in range(len(table)):
    while s and s[-1][0] < table[i]:
        v, idx = s.pop()
        result[idx] = table[i]
    else:
        s.append((table[i], i))
print(" ".join(map(str, result)))
